jupyterlite_xeus_python/env_build_addon.py

```diff
@@ -0,0 +1,219 @@
+)
+from tempfile import TemporaryDirectory
+import shutil
+from pathlib import Path
+
+from traitlets import List
+
+from emboa.file_packager import pack_python_core
+
+from jupyterlite.constants import SHARE_LABEXTENSIONS
+from jupyterlite.addons.federated_extensions import FederatedExtensionAddon
+
+PYTHON_VERSION = "3.10"
+
+CHANNELS = [
+    "https://repo.mamba.pm/emscripten-forge",
+    "https://repo.mamba.pm/conda-forge"
+]
+PLATFORM = "emscripten-32"
+
+SILENT = dict(stdout=DEVNULL, stderr=DEVNULL)
+
+try:
+    from mamba.api import create as mamba_create
+    MAMBA_PYTHON_AVAILABLE = True
+except ImportError:
+    MAMBA_PYTHON_AVAILABLE = False
+
+try:
+    check_call(["mamba", "--version"], **SILENT)
+    MAMBA_AVAILABLE = True
+except FileNotFoundError:
+    MAMBA_AVAILABLE = False
+
+try:
+    check_call(["micromamba", "--version"], **SILENT)
+    MICROMAMBA_AVAILABLE = True
+except FileNotFoundError:
+    MICROMAMBA_AVAILABLE = False
+
+try:
+    check_call(["conda", "--version"], **SILENT)
+    CONDA_AVAILABLE = True
+except FileNotFoundError:
+    CONDA_AVAILABLE = False
+
+
+class PackagesList(List):
+    def from_string(self, s):
+        return s.split(",")
+
+
+class XeusPythonEnv(FederatedExtensionAddon):
+
+    __all__ = ["pre_build", "post_build"]
+
+    packages = PackagesList([]).tag(
+        config=True,
+        description="A comma-separated list of packages to install in the xeus-python env"
+    )
+
+    @property
+    def specs(self):
+        """The package specs to install in the environment."""
+        return [f"python={PYTHON_VERSION}", "xeus-python", *self.packages]
+
+    @property
+    def prefix_path(self):
+        """The environment prefix."""
+        return Path(self.root_prefix) / "envs" / self.env_name
+
+    def __init__(self, *args, **kwargs):
+        super(XeusPythonEnv, self).__init__(*args, **kwargs)
+
+        self.cwd = TemporaryDirectory()
+        self.root_prefix = "/tmp/xeus-python-kernel"
+        self.env_name = "xeus-python-kernel"
+
+    def pre_build(self, manager):
+        """yield a doit task to create the emscripten-32 env and grab anything we need from it"""
+        # Bail early if there is nothing to do
+        if not self.packages:
+            return []
+
+        # Cleanup tmp dir
+        shutil.rmtree(self.root_prefix, ignore_errors=True)
+
+        # Create emscripten env with the given packages
+        self.create_env()
+
+        # Pack the environment
+        pack_python_core(
+            self.prefix_path,
+            outname=Path(self.cwd.name) / "python_data",
+            version=PYTHON_VERSION,
+            export_name="globalThis.Module",
+        )
+
+        # Find the federated extensions in the emscripten-env and install them
+        root = self.prefix_path / SHARE_LABEXTENSIONS
+
+        if not self.is_sys_prefix_ignored():
+            for pkg_json in self.env_extensions(root):
+                yield from self.copy_one_extension(pkg_json)
```

Comment:
I guess a sensible approach for the xeus-python addon would be to ignore that flag? (I'll check if it is ignored)

A typical user would probably always want to have the labextensions from the emscripten env be included? Otherwise they would not have installed them.